:memo: Creating a draft for upcoming cross-contract tutorial (#857)
o1-labs/docs2: docs/zkapps/tutorials/12-cross-contract-calls.mdx

---
title: 'Tutorial 12: Cross Contract Calls'
hide_title: true
sidebar_label: 'Tutorial 12: Cross Contract Calls'
description: Guided steps to learn about the ways to interact with a smart contract from another smart contract.
keywords:
  - smart contracts
  - zkapps
  - zero knowledge proof programming
  - zk proof
  - zk
  - cross contract call
  - mina
---

:::info

Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.

:::

# Tutorial 12: Cross Contract Calls

In this tutorial, you learn how smart contracts on a blockchain can interact by calling functions in each other's code, enabling building modular and complex decentralized applications.

> Cross contract calls allow smart contracts on a blockchain to interact with each other. This enables building complex decentralized applications (dApps) from multiple modular components. In a cross contract call, a function in one smart contract can call a function in another smart contract to leverage existing code and functionality.

Reviewer: ```suggestion
Cross contract calls allow smart contracts on a blockchain to interact with each other. This enables the building of complex decentralized applications (Dapps) from multiple modular components. In a cross-contract call, a function in one smart contract can call a function in another smart contract to leverage existing code and functionality.
```

Reviewer: use Dapp for the capitalization (yes, it's different than others) https://github.com/o1-labs/docs2/wiki/Word-list#dapp

> Cross contract calls allow smart contracts on a blockchain to interact with each other. This enables the building of complex decentralized applications (Dapps) from multiple modular components. In a cross-contract call, a function in one smart contract can call a function in another smart contract to leverage existing code and functionality.

Author: I was thinking the same !!
I was thinking of adding a image but an use case example sounds awesome.
Will check and add it in couple of days.

Thank You 🙌🏻

This tutorial demonstrates passing data between contracts, handling events, and returning values when contracts call each other.

> The full example code is provided in the [12-cross-contract-calls/src/](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/12-cross-contract-calls/src) example files.

Reviewer: where are the example files? are they in this PR?
let's put the example files mentioned in the tutorial in a new folder here https://github.com/o1-labs/docs2/tree/main/examples/zkapps

Author: Once this PR get's merges the link will start working.

Reviewer: I didn't recognize all three smart contracts as being present, since in a single file. Thanks @iregina  for testing the code!

> In this tutorial you will learn about passing data between contracts, handling events and return values when contracts call each other. 
> 
> The full example code is provided in the [12-cross-contract-calls/src/](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/12-cross-contract-calls/src) example files.

Reviewer: ```suggestion
This tutorial demonstrates passing data between contracts, handling events, and returning values when contracts call each other. 

The full example code is provided in the [12-cross-contract-calls/src/](https://github.com/o1-labs/docs2/tree/main/examples/zkapps/12-cross-contract-calls/src) example files.
```

## Prerequisites

- Make sure you have the latest version of the zkApp CLI installed:

    ```sh
    $ npm install -g zkapp-cli
    ```

- Ensure your environment meets the [Prerequisites](/zkapps/tutorials#prerequisites) for zkApp Developer Tutorials.

This tutorial has been tested with:

> - [Mina zkApp CLI](https://github.com/o1-labs/zkapp-cli) version `0.16.0`

Reviewer: ```suggestion
- [zkApp CLI](https://github.com/o1-labs/zkapp-cli) version `0.16.0`
```

## Write the ZkProgram

Now, the fun part! Write your smart contract in the `src/Composability.ts` file.

A final version of the smart contract is provided in the [Composability.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/12-cross-contract-calls/src/Composability.ts) example file.

### Copy the example

Use the existing code in the [Composability.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/12-cross-contract-calls/src/Composability.ts) example file.

1. First, open the [Composability.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/12-cross-contract-calls/src/Composability.ts) example file.
	
1. Copy the file's entire contents into your project `src/Composability.ts` file.

### Imports and Incrementer smart contract 

First, bring in imports and set up the first smart contract `Incrementer`.

```typescript
import {
  Field,
  method,
  Mina,
  AccountUpdate,
  PrivateKey,
  SmartContract,
  state,
  State,
} from 'o1js';

class Incrementer extends SmartContract {
  @method increment(x: Field): Field {
    return x.add(1);
  }
}

```

> This Incrementer contract will add 1 to the Field argument which is passed.

Reviewer: ```suggestion
This Incrementer contract adds `1` to the `Field` argument, which is passed.
```

### Adder smart contract 

> Let's bring in the second contract `Adder` which will return the addtion of two numbers and add 1 to the result.
> The addition of 1 to the result is outsourced to the Incrementer smart contract by creating a new object by passing it's address.

Reviewer: ```suggestion
Now bring in the second contract `Adder` that returns the addition of two numbers and adds `1` to the result.
The addition of `1` to the result is outsourced to the `Incrementer` smart contract by creating a new object by passing its address.
```

```typescript
class Adder extends SmartContract {
  @method addPlus1(x: Field, y: Field): Field {
    let sum = x.add(y);
    let incrementer = new Incrementer(incrementerAddress);
    return incrementer.increment(sum);
  }
}

```

### Caller smart contract 

> The final smart contract `Caller` will call the `addPlus1` method of `Adder` smart contract and emit the stored result return by it.

Reviewer: ```suggestion
The final smart contract `Caller` calls the `addPlus1()` method of the `Adder` smart contract and emits the stored result that is returned.
```

    let sum = adder.addPlus1(x, y);
    this.emitEvent('sum', sum);
    this.sum.set(sum);
  }
}

```

The code to interact with the smart contract:

```typescript
const doProofs = true;

let Local = Mina.LocalBlockchain({ proofsEnabled: doProofs });
Mina.setActiveInstance(Local);

let feePayerKey = Local.testAccounts[0].privateKey;
let feePayer = Local.testAccounts[0].publicKey;

let incrementerKey = PrivateKey.random();
let incrementerAddress = incrementerKey.toPublicKey();

let adderKey = PrivateKey.random();
let adderAddress = adderKey.toPublicKey();

let zkappKey = PrivateKey.random();
let zkappAddress = zkappKey.toPublicKey();

let zkapp = new Caller(zkappAddress);
let adderZkapp = new Adder(adderAddress);
let incrementerZkapp = new Incrementer(incrementerAddress);

if (doProofs) {
  console.log('compile (incrementer)');
  await Incrementer.compile();
  console.log('compile (adder)');
  await Adder.compile();
  console.log('compile (caller)');
  await Caller.compile();
}

console.log('deploy');
let tx = await Mina.transaction(feePayer, () => {
  AccountUpdate.fundNewAccount(feePayer, 3);
  zkapp.deploy();
  adderZkapp.deploy();
  incrementerZkapp.deploy();
});

> await tx.sign([feePayerKey, zkappKey, adderKey, incrementerKey]).send();

Reviewer 2: @LuffySama-Dev Can you also leave a comment about what is happening here?

Author: Hi @iregina ,
Yes sure. I will add a comment.

Thank You ☺️

Reviewer: I like your question about comments @LuffySama-Dev  
Helpful comments in the code is the pathway to sharing knowledge 

Please add descriptive comments

Author: Hi @barriebyron , @iregina ,

I have committed the changes, can you please help to check and let me know if any changes are required ?
- Also, I am planning on adding a flow diagram of how call is going from one contract to other an so.
What do you think ?

Thank You 😃

console.log('call interaction');
tx = await Mina.transaction(feePayer, () => {
  zkapp.callAddAndEmit(Field(5), Field(6));
});
console.log('proving (3 proofs.. can take a bit!)');
await tx.prove();
console.log(tx.toPretty());
await tx.sign([feePayerKey]).send();

console.log('state: ' + zkapp.sum.get());

``` 

> In this example, we are spinning up the Mina chain locally and deploying all three smart contracts locally.
> 
> Then we are calling `callAddAndEmit` method from `Caller` smart contract which will take two numbers as arguments, then call the `Adder` smart contract which will add these two numbers and pass the result in `Incrementer` smart contract which will increment the result by 1.
> 
> In our case we should see state equal to 12.

Reviewer: ```suggestion
In this example, you spin up the Mina chain and deploy all three smart contracts locally.

Then you call the `callAddAndEmit` method from the `Caller` smart contract, which takes two numbers as arguments, then call the `Adder` smart contract, which adds these two numbers and passes the result in the `Incrementer` smart contract, which increments the result by 1.

When run successfully, the state is equal to 12.
```
